homeassistant/components/kodi/media_player.py

```diff
@@ -794,3 +852,206 @@ def _find(key_word, words):
             out[i][1] = rate
 
         return sorted(out, key=lambda out: out[1], reverse=True)
+
+    async def async_browse_media(self, media_content_type=None, media_content_id=None):
+        """Implement the websocket media browsing helper."""
+        if media_content_type in [None, "library"]:
+            return await self.hass.async_add_executor_job(library_payload, self._kodi)
+
+        payload = {
+            "search_type": media_content_type,
+            "search_id": media_content_id,
+        }
+        response = await build_item_response(self._kodi, payload)
+        if response is None:
+            raise BrowseError(
+                f"Media not found: {media_content_type} / {media_content_id}"
+            )
+        return response
+
+
+async def build_item_response(media_library, payload):
```

Comment:
Should we break everything from here on down to a separate file?

Reply:
Also, can we DRY things up with the existing search functions?

Reply:
I've split it up. It should become a lot simpler when we change the upstream module.